net/golioth/coap_req.c

```diff
@@ -0,0 +1,709 @@
+	struct golioth_coap_req *req;
+	int err;
+
+	err = golioth_coap_req_new(&req, client, method, COAP_TYPE_CON,
+				   GOLIOTH_COAP_MAX_NON_PAYLOAD_LEN + path_len + data_len,
+				   cb, user_data);
+	if (err) {
+		LOG_ERR("Failed to create new CoAP GET request: %d", err);
+		goto free_req;
+	}
+
+	if (method == COAP_METHOD_GET && (flags & GOLIOTH_COAP_REQ_OBSERVE)) {
+		req->is_observe = true;
+		req->is_pending = true;
+
+		err = coap_append_option_int(&req->request, COAP_OPTION_OBSERVE, 0 /* register */);
+		if (err) {
+			LOG_ERR("Unable add observe option");
+			goto free_req;
+		}
+	}
+
+	err = coap_packet_append_uri_path_from_pathv(&req->request, pathv);
+	if (err) {
+		LOG_ERR("Unable add uri path to packet");
+		goto free_req;
+	}
+
+	if (method != COAP_METHOD_GET && method != COAP_METHOD_DELETE) {
+		err = coap_append_option_int(&req->request, COAP_OPTION_CONTENT_FORMAT, format);
+		if (err) {
+			LOG_ERR("Unable add content format to packet");
+			goto free_req;
+		}
+	}
+
+	if (!(flags & GOLIOTH_COAP_REQ_NO_RESP_BODY)) {
+		err = coap_append_option_int(&req->request, COAP_OPTION_ACCEPT, format);
+		if (err) {
+			LOG_ERR("Unable add content format to packet");
+			goto free_req;
+		}
+	}
+
+	if (data && data_len) {
+		err = coap_packet_append_payload_marker(&req->request);
+		if (err) {
+			LOG_ERR("Unable add payload marker to packet");
+			goto free_req;
+		}
+
+		err = coap_packet_append_payload(&req->request, data, data_len);
+		if (err) {
+			LOG_ERR("Unable add payload to packet");
+			goto free_req;
+		}
+	}
+
+	return golioth_coap_req_schedule(req);
+
+free_req:
+	golioth_coap_req_free(req);
+
+	return err;
+}
+
+struct golioth_req_sync_data {
+	struct k_sem sem;
+	int err;
+
+	golioth_req_cb_t cb;
+	void *user_data;
+};
+
+static int golioth_req_sync_cb(struct golioth_req_rsp *rsp)
+{
+	struct golioth_req_sync_data *sync_data = rsp->user_data;
+	int err = 0;
+
+	if (rsp->err) {
+		sync_data->err = rsp->err;
+		goto finish_sync_call;
+	}
+
+	if (sync_data->cb) {
+		rsp->user_data = sync_data->user_data;
+
+		err = sync_data->cb(rsp);
+		if (err) {
+			goto finish_sync_call;
+		}
+	}
+
+	if (rsp->get_next) {
+		rsp->get_next(rsp->get_next_data, 0);
+		return 0;
+	}
+
+finish_sync_call:
+	k_sem_give(&sync_data->sem);
+
+	return err;
+}
+
+int golioth_coap_req_sync(struct golioth_client *client,
+			  enum coap_method method,
+			  const uint8_t **pathv,
+			  enum golioth_content_format format,
+			  const uint8_t *data, size_t data_len,
+			  golioth_req_cb_t cb, void *user_data,
+			  int flags)
+{
+	struct golioth_req_sync_data sync_data = {
+		.cb = cb,
+		.user_data = user_data,
+	};
+	int err;
+
+	k_sem_init(&sync_data.sem, 0, 1);
+
+	err = golioth_coap_req_cb(client, method, pathv, format,
+				  data, data_len,
+				  golioth_req_sync_cb, &sync_data,
+				  flags);
+	if (err) {
+		LOG_WRN("Failed to make CoAP request: %d", err);
+		return err;
+	}
+
+	k_sem_take(&sync_data.sem, K_FOREVER);
```

Comment:
Is it ever possible for `sync_data.sem` to never be given? Is it guaranteed to always be given?

Reply:
It is assumed so. However I realized that this is not true in case of connection close. In case of connection restart "system client" loop should continue resending / expiring pending packets with this implementation. This is bad (because packets from previous connection should not be "resent" with new connection), however it would unblock application thread at some point.

Added https://golioth.atlassian.net/browse/DSDK-390 to cover that.

Reply:
As a defensive measure, you could add an assert after a long time, in case it can't be taken for any reason. Something like:

```c
assert(k_sem_take(&sync_data.sem, K_SECONDS(100)) == 0);
```